streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultStateConstrainedBalancedAssignor.java

```diff
@@ -0,0 +1,218 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DefaultStateConstrainedBalancedAssignor implements StateConstrainedBalancedAssignor {
+
+    private final Map<String, List<TaskId>> previousAssignment;
+
+    public DefaultStateConstrainedBalancedAssignor(final Map<String, List<TaskId>> previousAssignment) {
+        this.previousAssignment = previousAssignment;
+    }
+
+    /**
+     * This assignment algorithm guarantees that all task for which caught-up clients exist are assigned to one of the
+     * caught-up clients. Tasks for which no caught-up client exist are assigned best-effort to satisfy the balance
+     * factor. There is not guarantee that the balance factor is satisfied.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @param balanceFactor balance factor (at least 1)
+     * @return assignment
+     */
+    @Override
+    public Map<String, List<TaskId>> assign(final Map<TaskId, List<ClientIdAndLag>> statefulTasksToRankedClients,
+                                            final int balanceFactor) {
+        final Map<String, List<TaskId>> assignment = initAssignment(statefulTasksToRankedClients);
+        final Map<TaskId, List<String>> tasksToCaughtUpClients = tasksToCaughtUpClients(statefulTasksToRankedClients);
+        final List<TaskId> unassignedTasksWithoutCaughtUpClients = new ArrayList<>();
+        assignTasksWithCaughtUpClients(
+            assignment,
+            unassignedTasksWithoutCaughtUpClients,
+            tasksToCaughtUpClients
+        );
+        assignTasksWithoutCaughtUpClients(
+            assignment,
+            unassignedTasksWithoutCaughtUpClients,
+            statefulTasksToRankedClients
+        );
+        balance(assignment, balanceFactor, statefulTasksToRankedClients, tasksToCaughtUpClients);
+        return assignment;
+    }
+
+    /**
+     * Initialises the assignment with the clients in the ranked clients map and assigns to each client an empty list.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @return initialised assignment with empty lists
+     */
+    private Map<String, List<TaskId>> initAssignment(final Map<TaskId, List<ClientIdAndLag>> statefulTasksToRankedClients) {
+        final Map<String, List<TaskId>> assignment = new HashMap<>();
+        final List<ClientIdAndLag> clientIdAndLags = statefulTasksToRankedClients.values().stream()
+            .findFirst().orElseThrow(() -> new IllegalStateException("The list of clients and lags must not be empty"));
+        for (final ClientIdAndLag clientIdAndLag : clientIdAndLags) {
+            assignment.put(clientIdAndLag.clientId(), new ArrayList<>());
+        }
+        return assignment;
+    }
+
+    /**
+     * Maps tasks to clients with caught-up states for the task.
+     *
+     * @param statefulTasksToRankedClients ranked clients map
+     * @return map from task IDs to lists of caught-up clients
```

Comment:
```suggestion
     * @return map from tasks with caught-up clients to the list of client candidates
```
or something similar to make it clear the map only contains tasks with caught-up clients